src/compiler/transformers/declarations.ts

```diff
@@ -282,6 +296,7 @@ export function transformDeclarations(context: TransformationContext) {
     const resolver = context.getEmitResolver();
     const options = context.getCompilerOptions();
     const { noResolve, stripInternal } = options;
+    const strictNullChecks = getStrictOptionValue(options, "strictNullChecks");

```

Comment:
Feels weird to have this detail in the transform; right now the only options that are used are ones explicitly related to declaration emit (noResolve/stripInternal).

My impression is that isolatedDeclarations is not assumed to know the compiler options; is the case that we can treat strictNullChecks as true for emit with the assumption that `undefined`/`null` are just going to be ignored by non-strict users?

Reply:
There is a bit of `strictNullChecks` that needs to be taken into account. This is used here in the context of inferring optional property declaration types. If `strictNullChecks` is on we need to add `undefined` to the type otherwise we don't. Without taking this into account any local inference optional declarations will not be possible:

```ts
class X { 
	// x?: 1 | 2 | undefined; with !strictNullChecks
	// x?: 1 | 2; // with !strictNullChecks
	x? = 1 as 1 | 2 
}
```
[Playground Link](https://www.typescriptlang.org/play?#code/MYGwhgzhAEAa0G9oFgBQBIA9J6APA-AFzQCM0APtAEwXQCuAdgCYCmAZgJYMtMDc0Adw4AXABbQAhBGEAnDsGEA5OiBABhUS2ABrCGiw4CxMpSr9sgkeKmz5Sles069GAtAC8paJC+mUqAF80IA)

Reply:
Sure but if a user doesn't have strictNullChecks enabled, aren't the two equivalent? Since the way that strictNullChecks=false works is to effectively treat null/undefined as though they don't exist.

Reply:
They are usually, but if you take the declaration from a project with `!strictNullChecks` and use them in a project `exactOptionalPropertyTypes` then the missing `| undefined` will make a difference.

Reply:
Thinking about it, it might actually be an improvement. Because a `!strictNullChecks` project should be fine with a `undefined`. 

I can change this to always add `| undefined` but the second problem is that it will make emit different between the case when we take the type from the checker (which does not add undefined) and when we try to get the type from the assertion. The solution would be to add the `|undefined` always form the checker as well. 

Maybe this should be a different PR though? One that always adds `|undefined` to optional properties even if `!strictNullChecks`

Reply:
> One that always adds `|undefined` to optional properties

` | undefined` is not implied by optional properties under `exactOptionalPropertyTypes`, so this would be unsafe

Reply:
I guess I see what you mean, in that this PR is only trying to touch assertions and therefore making this code unconditional would not change any other declarations?

Reply:
Exactly. Even it's final form, `!isolatedDeclarations` some types will still come from the checker, I don't think adding undefined just sometimes would be a good thing.